### ISSUE TITLE
inconsistency in pending accounting

### ISSUE THREAD

**Issue by vlimant** [dmwm/WMCore#8256]
while trying to get why

https://cmsweb.cern.ch/couchdb/workqueue/_design/WorkQueue/_rewrite/elementsInfo?request=pdmvserv_task_HGC-PhaseIITDRFall17DR-00039__v1_T_171006_181519_8352

is not getting acquired by any of the available agent in the production team, I was unable to reconcile

cmst1@vocms0253:/data/srv/wmagent/current $ $manage execute-agent wmagent-resource-control --site-name=T1_RU_JINR -p
Executing wmagent-resource-control --site-name=T1_RU_JINR -p ...
Thresholds and current status for all sites:

T1_RU_JINR - 109 running, 602 pending, 3613 running slots total, 602 pending slots total, Site is Normal:
  Merge - 0 running, 0 pending, 100 max running, 25 max pending, priority 5
  Cleanup - 0 running, 0 pending, 100 max running, 25 max pending, priority 5
  Skim - 0 running, 0 pending, 100 max running, 25 max pending, priority 3
  LogCollect - 0 running, 0 pending, 100 max running, 25 max pending, priority 3
  Processing - 103 running, 541 pending, 3613 max running, 541 max pending, priority 0
    44 - /pdmvserv_task_SMP-PhaseIITDRFall17DR-00002__v1_T_171006_172941_6056/SMP-PhaseIITDRFall17DR-00002_0
    2 - /pdmvserv_task_SUS-PhaseIITDRFall17wmLHEGS-00022__v1_T_171010_161142_9822/SUS-PhaseIITDRFall17wmLHEGS-00022_0/SUS-PhaseIITDRFall17wmLHEGS-00022_0MergeRAWSIMoutput/SUS-PhaseIITDRFall17DR-00043_0
    1 - /pdmvserv_task_SMP-PhaseIITDRFall17wmLHEGS-00006__v1_T_171008_201204_4047/SMP-PhaseIITDRFall17wmLHEGS-00006_0/SMP-PhaseIITDRFall17wmLHEGS-00006_0MergeRAWSIMoutput/SMP-PhaseIITDRFall17DR-00013_0
    10 - /pdmvserv_task_SMP-PhaseIITDRFall17wmLHEGS-00020__v1_T_171012_001827_7735/SMP-PhaseIITDRFall17wmLHEGS-00020_0/SMP-PhaseIITDRFall17wmLHEGS-00020_0MergeRAWSIMoutput/SMP-PhaseIITDRFall17DR-00039_0
    5 - /pdmvserv_task_SMP-PhaseIITDRFall17DR-00007__v1_T_171006_181121_5757/SMP-PhaseIITDRFall17DR-00007_0/SMP-PhaseIITDRFall17DR-00007_1
    1 - /pdmvserv_task_SMP-PhaseIITDRFall17wmLHEGS-00011__v1_T_171012_001731_6164/SMP-PhaseIITDRFall17wmLHEGS-00011_0/SMP-PhaseIITDRFall17wmLHEGS-00011_0MergeRAWSIMoutput/SMP-PhaseIITDRFall17DR-00023_0/SMP-PhaseIITDRFall17DR-00023_1
    2 - /pdmvserv_task_SMP-PhaseIITDRFall17wmLHEGS-00012__v1_T_171012_001816_7413/SMP-PhaseIITDRFall17wmLHEGS-00012_0/SMP-PhaseIITDRFall17wmLHEGS-00012_0MergeRAWSIMoutput/SMP-PhaseIITDRFall17DR-00024_0/SMP-PhaseIITDRFall17DR-00024_1
    1 - /mcremone_ACDC0_task_HGC-PhaseIITDRFall17DR-00023__v1_T_171017_121011_1707/HGC-PhaseIITDRFall17DR-00023_1
    5 - /pdmvserv_task_SMP-PhaseIITDRFall17wmLHEGS-00012__v1_T_171012_001816_7413/SMP-PhaseIITDRFall17wmLHEGS-00012_0/SMP-PhaseIITDRFall17wmLHEGS-00012_0MergeRAWSIMoutput/SMP-PhaseIITDRFall17DR-00025_0/SMP-PhaseIITDRFall17DR-00025_1
    433 - /pdmvserv_task_SMP-PhaseIITDRFall17DR-00007__v1_T_171006_181121_5757/SMP-PhaseIITDRFall17DR-00007_0
    1 - /pdmvserv_task_SMP-PhaseIITDRFall17wmLHEGS-00012__v1_T_171012_001816_7413/SMP-PhaseIITDRFall17wmLHEGS-00012_0/SMP-PhaseIITDRFall17wmLHEGS-00012_0MergeRAWSIMoutput/SMP-PhaseIITDRFall17DR-00024_0
    133 - /pdmvserv_task_SMP-PhaseIITDRFall17DR-00002__v1_T_171006_172941_6056/SMP-PhaseIITDRFall17DR-00002_0/SMP-PhaseIITDRFall17DR-00002_1
    2 - /pdmvserv_task_HIG-PhaseIITDRFall17wmLHEGS-00015__v1_T_171010_201726_6880/HIG-PhaseIITDRFall17wmLHEGS-00015_0/HIG-PhaseIITDRFall17wmLHEGS-00015_0MergeRAWSIMoutput/HIG-PhaseIITDRFall17DR-00029_0
    4 - /pdmvserv_task_SUS-PhaseIITDRFall17wmLHEGS-00026__v1_T_171010_161216_4884/SUS-PhaseIITDRFall17wmLHEGS-00026_0/SUS-PhaseIITDRFall17wmLHEGS-00026_0MergeRAWSIMoutput/SUS-PhaseIITDRFall17DR-00051_0
  Production - 6 running, 61 pending, 3613 max running, 541 max pending, priority 0
    4 - /pdmvserv_task_HIG-RunIISummer15wmLHEGS-01737__v1_T_171012_121755_352/HIG-RunIISummer15wmLHEGS-01737_0
    1 - /pdmvserv_task_HIG-RunIIFall17GS-00004__v1_T_171012_121336_4544/HIG-RunIIFall17GS-00004_0
    55 - /pdmvserv_task_HIG-PhaseIITDRFall17wmLHEGS-00015__v1_T_171010_201726_6880/HIG-PhaseIITDRFall17wmLHEGS-00015_0
    6 - /pdmvserv_task_HIG-RunIIFall17GS-00002__v1_T_171012_121333_7391/HIG-RunIIFall17GS-00002_0
    1 - /pdmvserv_task_EGM-RunIIFall17GS-00005__v1_T_171014_121411_2635/EGM-RunIIFall17GS-00005_0
  Harvesting - 0 running, 0 pending, 100 max running, 25 max pending, priority 3

and

cmst1@vocms0253:/data/srv/wmagent/current $ condor_q -const 'WMAgent_RequestName =="vlimant_task_TOP-RunIIFall17wmLHEGS-00010__v1_T_171017_103100_8509"' -af JobStatus ExtDESIRED_Sites | sort | uniq -c
   4135 1 T1_DE_KIT,T1_ES_PIC,T1_FR_CCIN2P3,T1_IT_CNAF,T1_RU_JINR,T1_UK_RAL,T1_US_FNAL,T2_BE_IIHE,T2_BE_UCL,T2_CH_CERN,T2_CH_CERN_HLT,T2_CN_Beijing,T2_DE_DESY,T2_DE_RWTH,T2_ES_CIEMAT,T2_FR_CCIN2P3,T2_FR_GRIF_IRFU,T2_FR_GRIF_LLR,T2_IN_TIFR,T2_IT_Bari,T2_IT_Legnaro,T2_IT_Pisa,T2_PT_NCG_Lisbon,T2_RU_JINR,T2_TW_NCHC,T2_UK_London_Brunel,T2_UK_London_IC,T2_UK_SGrid_RALPP,T2_US_Caltech,T2_US_Florida,T2_US_MIT,T2_US_Nebraska,T2_US_Purdue,T2_US_UCSD,T2_US_Vanderbilt,T2_US_Wisconsin
      1 1 T2_US_Florida
     20 2 T1_DE_KIT,T1_ES_PIC,T1_FR_CCIN2P3,T1_IT_CNAF,T1_RU_JINR,T1_UK_RAL,T1_US_FNAL,T2_BE_IIHE,T2_BE_UCL,T2_CH_CERN,T2_CH_CERN_HLT,T2_CN_Beijing,T2_DE_DESY,T2_DE_RWTH,T2_ES_CIEMAT,T2_FR_CCIN2P3,T2_FR_GRIF_IRFU,T2_FR_GRIF_LLR,T2_IN_TIFR,T2_IT_Bari,T2_IT_Legnaro,T2_IT_Pisa,T2_PT_NCG_Lisbon,T2_RU_JINR,T2_TW_NCHC,T2_UK_London_Brunel,T2_UK_London_IC,T2_UK_SGrid_RALPP,T2_US_Caltech,T2_US_Florida,T2_US_MIT,T2_US_Nebraska,T2_US_Purdue,T2_US_UCSD,T2_US_Vanderbilt,T2_US_Wisconsin
      1 2 T2_CH_CERN

4135 job pending to JINR, created in the last couple of hours, while WQE like

https://cmsweb.cern.ch/couchdb/workqueue/_design/WorkQueue/_rewrite/element/0e3ef548cd7dd6e9a1c03c1fd1576ce3
 
is out there to be ran at high priority.
These PhaseII are quite the urgent matter, hence the label

**Comment by ticoann** [comment]
I see a few of jobs are acquired in vocms0250, 0253. I will check whether there is problem to prevent acquiring more jobs.

**Comment by ticoann** [comment]
I checked 3 production agents currently not in drain, all the agents' resource are full for T1_RU_JINR

```
cmst1@vocms0250:/data/srv/wmagent/current $ $manage execute-agent check-request-wq-status -r pdmvserv_task_HGC-PhaseIITDRFall17DR-00039__v1_T_171006_181519_8352 -s Available
Executing check-request-wq-status -r pdmvserv_task_HGC-PhaseIITDRFall17DR-00039__v1_T_171006_181519_8352 -s Available ...
### GQ at https://cmsweb.cern.ch/couchdb/workqueue: There are a total of 24 elements in Available
### summary for pdmvserv_task_HGC-PhaseIITDRFall17DR-00039__v1_T_171006_181519_8352 ###
  Priority: 110000, available elements: 24 
  Possible sites to run: [u'T1_RU_JINR']
  Higher priority elements: 0, total available jobs: 0
All pre condition checks are passed to pullWork to LQ
Possible sites for the request pdmvserv_task_HGC-PhaseIITDRFall17DR-00039__v1_T_171006_181519_8352
T1_RU_JINR: {'priority': {110000: 409.0}, 'total jobs created': 409.0, 'slots': 301.0}
WARNING:root:No available work in parent queue.
Get available work from GQ
Resource is full no work will be pulled to LQ


cmst1@vocms0251:/data/srv/wmagent/current $ $manage execute-agent check-request-wq-status -r pdmvserv_task_HGC-PhaseIITDRFall17DR-00039__v1_T_171006_181519_8352 -s Available
Executing check-request-wq-status -r pdmvserv_task_HGC-PhaseIITDRFall17DR-00039__v1_T_171006_181519_8352 -s Available ...
### GQ at https://cmsweb.cern.ch/couchdb/workqueue: There are a total of 24 elements in Available
### summary for pdmvserv_task_HGC-PhaseIITDRFall17DR-00039__v1_T_171006_181519_8352 ###
  Priority: 110000, available elements: 24 
  Possible sites to run: [u'T1_RU_JINR']
  Higher priority elements: 0, total available jobs: 0
All pre condition checks are passed to pullWork to LQ
Possible sites for the request pdmvserv_task_HGC-PhaseIITDRFall17DR-00039__v1_T_171006_181519_8352
T1_RU_JINR: {'priority': {110000: 1617.0}, 'total jobs created': 1617.0, 'slots': 301.0}
WARNING:root:No available work in parent queue.
Get available work from GQ
Resource is full no work will be pulled to LQ

cmst1@vocms0253:/data/srv/wmagent/current $ $manage execute-agent check-request-wq-status -r pdmvserv_task_HGC-PhaseIITDRFall17DR-00039__v1_T_171006_181519_8352 -s Available
Executing check-request-wq-status -r pdmvserv_task_HGC-PhaseIITDRFall17DR-00039__v1_T_171006_181519_8352 -s Available ...
### GQ at https://cmsweb.cern.ch/couchdb/workqueue: There are a total of 24 elements in Available
### summary for pdmvserv_task_HGC-PhaseIITDRFall17DR-00039__v1_T_171006_181519_8352 ###
  Priority: 110000, available elements: 24 
  Possible sites to run: [u'T1_RU_JINR']
  Higher priority elements: 0, total available jobs: 0
All pre condition checks are passed to pullWork to LQ
Possible sites for the request pdmvserv_task_HGC-PhaseIITDRFall17DR-00039__v1_T_171006_181519_8352
T1_RU_JINR: {'priority': {110000: 334.0}, 'total jobs created': 334.0, 'slots': 301.0}
WARNING:root:No available work in parent queue.
Get available work from GQ
Resource is full no work will be pulled to LQ
```

**Comment by vlimant** [comment]
hi @ticoann  thanks for taking a look. that was my understanding too earlier while bugging @amaltaro . but then the same agents that are indeed full for JINR did acquire some other lower priority work for wqe that can and will run at JINR.

furthermore, from the initial text

T1_RU_JINR - 109 running, 602 pending, 3613 running slots total, 602 pending slots total, Site is Normal:
while
4135 1 T1_DE_KIT,T1_ES_PIC,T1_FR_CCIN2P3,T1_IT_CNAF,T1_RU_JINR,T

so schedd says 4135 pending, Agent sees 602 only. Unless I am mistaken.

so why is the agent able to pull work that will go to JINR, while the work aimed at JINR is not pulled ?

**Comment by ticoann** [comment]
Hi Jean-Roch,
> 4135 1 T1_DE_KIT,T1_ES_PIC,T1_FR_CCIN2P3,T1_IT_CNAF,T1_RU_JINR,T

How did you get this number? Is it only for T1_RU_JINR or combined with other sites? 

> but then the same agents that are indeed full for JINR did acquire some other lower priority work for wqe that can and will run at JINR

You mean from GQ? Hmm, that shouldn't happen if higher priority is there at the same time. Actually, Alan was telling me there might be a bug causing that. I will check on that.

**Comment by vlimant** [comment]
4135 is from condor_q as provided initially

cmst1@vocms0253:/data/srv/wmagent/current $ condor_q -const 'WMAgent_RequestName =="vlimant_task_TOP-RunIIFall17wmLHEGS-00010__v1_T_171017_103100_8509"' -af JobStatus ExtDESIRED_Sites | sort | uniq -c
4135 1 T1_DE_KIT,T1_ES_PIC,T1_FR_CCIN2P3,T1_IT_CNAF,T1_RU_JINR,T1_

**Comment by amaltaro** [comment]
I think I understand the numbers provided by Seangchan
`T1_RU_JINR: {'priority': {110000: 409.0}, 'total jobs created': 409.0, 'slots': 301.0}`

this takes the resources for JINR as seen by GQ -> LQ acquisition. Which means, anything <= 110k prio will not be acquired by the agent until there are slots available.

However, the 409 jobs created is not very accurate because jobs use late binding, so AFAIK it does a distribution of the jobs and the sitewhitelist (avg).

The condor pending threshold is set to 602 jobs, which is full in vocms0253.

And querying condor, for jobs pending and with u'T1_RU_JINR' in the DESIRED_Sites list, there are several thousands jobs with JINR in the whitelist with all sorts of priorities, worth pasting though:
```
cmst1@vocms0253:/data/srv/wmagent/current $ condor_q -const 'JobStatus =?= 1' -af JobPrio WMAgent_RequestName DESIRED_Sites | grep T1_RU_JINR | sort | uniq -c | sort -n -k 2
...
     96 85000 pdmvserv_task_HIG-PhaseIITDRFall17wmLHEGS-00015__v1_T_171010_201726_6880 T1_FR_CCIN2P3,T1_RU_JINR,T2_FR_CCIN2P3,T2_FR_GRIF_IRFU
     31 110000 pdmvserv_task_SMP-PhaseIITDRFall17DR-00002__v1_T_171006_172941_6056 T1_DE_KIT,T1_ES_PIC,T1_FR_CCIN2P3,T1_IT_CNAF,T1_RU_JINR,T1_UK_RAL,T2_UK_London_Brunel
    457 110000 pdmvserv_task_SMP-PhaseIITDRFall17DR-00007__v1_T_171006_181121_5757 T1_FR_CCIN2P3,T1_RU_JINR
    616 110000 pdmvserv_task_SMP-PhaseIITDRFall17DR-00002__v1_T_171006_172941_6056 T1_FR_CCIN2P3,T1_RU_JINR,T1_UK_RAL
     76 110000 pdmvserv_task_SMP-PhaseIITDRFall17DR-00002__v1_T_171006_172941_6056 T1_FR_CCIN2P3,T1_RU_JINR
      8 110000 pdmvserv_task_SMP-PhaseIITDRFall17DR-00002__v1_T_171006_172941_6056 T1_ES_PIC,T1_FR_CCIN2P3,T1_IT_CNAF,T1_RU_JINR
      9 110000 pdmvserv_task_SMP-PhaseIITDRFall17DR-00007__v1_T_171006_181121_5757 T1_ES_PIC,T1_FR_CCIN2P3,T1_IT_CNAF,T1_RU_JINR
     94 110000 pdmvserv_task_SMP-PhaseIITDRFall17DR-00002__v1_T_171006_172941_6056 T1_FR_CCIN2P3,T1_RU_JINR,T2_FR_CCIN2P3,T2_FR_GRIF_IRFU,T2_FR_GRIF_LLR
```

thus, > 1k jobs pending for JINR (and others) with prio 110k.

To summarize, numbers given by the agent (from wmbs) are not accurate due to late binding, nonetheless, I don't see any priority issues and the agent is doing what's supposed to do.

If the request is so damn urgent, why do we have to set its priority to the same as 100s of other requests...

IMO, we should implement some logic in reqmgr such that it bumps the request priority for every X days/weeks.

**Comment by amaltaro** [comment]
for completeness, agents vocms0250 and vocms0251 are on the same situation, there are > 1k pending jobs with JINR in the whitelisted (all prio 110k)

**Comment by vlimant** [comment]
the request is at the same priority as the other 110k priority. it's just that people are nervous it has not started yet. I want to be damn sure the agents are not stuck not pulling it down because of some bug in the logic.
if the agents keep on getting stuff that can run anywhere, and JINR, at lower priority then the threshold will always be at max and those WQE will never get in, which looks like a flaw.
but I want we are sure of this before taking actions.

**Comment by vlimant** [comment]
-1 to, "we should implement some logic in reqmgr such that it bumps the request priority for every X days/weeks."
making it 1M priority these particular WQE will not get them in, would it ?

**Comment by amaltaro** [comment]
making it 111k would be enough to get more elements acquired in the agent. Of course, there is still the condor pending thresholds, but any new free slots at JINR would pull these 111k jobs (unless there are merge/logcollect/cleanup for the same site OR higher prio)

**Comment by amaltaro** [comment]
Besides the wmbs numbers that are not accurate due to the late binding, there is nothing to be fixed here. I'm closing the issue.

**Comment by vlimant** [comment]
is there a chance at having the wmbs numbers more accurate ? the issue was essentially about that.
the request was cloned because there was just no agent pulling these wqe. 
So I say we have a problem, but you can decide to ignore it. It will come back I can tell you.

**Comment by amaltaro** [comment]
The only way to get it more accurate would be changing how we assign sites to pending jobs (with late binding), such that we assign a pending job to every single site in the DESIRED_Sites. Of course that would decrease the pending pressure, but nothing that we couldn't "fix" with an increased threshold.

About the agent not pulling those elements, all the debugging we have done shown no problems and the agent was doing its job.
